### PR TITLE
Test runner for batched gradient computation with vmap

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43664 Test runner for batched gradient computation with vmap**

This PR implements the test runner for batched gradient computation with
vmap. It also implements the batching rule for sigmoid_backward and
tests that one can compute batched gradients with sigmoid (and batched
2nd gradients).

Test Plan:
- New tests: `python test/test_vmap.py -v`

Differential Revision: [D23358555](https://our.internmc.facebook.com/intern/diff/D23358555)

### CODE REVIEW

File: test/test_vmap.py [hunk 1]
```diff
@@ -640,35 +643,42 @@ def randn(size, device='cpu', dtype=torch.float):
     def randp1(size, device='cpu', dtype=torch.float):
         return torch.rand(size, device=device, dtype=dtype) + 1
 
+# Tests vmap(op, in_dims, out_dims)(*inputs) by comparing the output to a
+# (slow) sequential map+stack fallback.
+#
+# check_view: Test if the first returned output is a view of the first input
+# check_propagates_grad: Test if the operation propagates gradients.
+def _vmap_test(self, op, inputs, in_dims=0, out_dims=0,
```
Comment (zou3519) [inline test/test_vmap.py:651]: This function was moved line-for-line from TestVmapOperators._vmap_test

File: test/test_vmap.py [hunk 2]
```diff
@@ -1150,5 +1160,100 @@ def test_no_random_op_support(self):
...
+    #       that should be differentiated.
+    # batch_size: the batch dim size for the batched grad
+    #
+    # NB: we only test computing batched gradients are in the second gradient
+    # computation. One specific use case that does this is computing the hessian
+    # matrix of a scalar-valued function; this is useful in Bayesian Logistic
+    # Regression.
+    # It might be useful to have a test that computes batched first gradients and
+    # then uses those to compute batched second gradients in the future.
+    def _batched_grad_grad_test(self, op, args, kwargs, output_process_fn=lambda x: x, batch_size=3):
```
Comment (zou3519) [inline test/test_vmap.py:1213]: One question is: I could make this (and _batched_grad_grad_test) a part of gradcheck, but I'm not sure if I should.

The nice thing about making it a part of gradcheck is that we can be sure that "batched gradient" works for literally all operators in PyTorch. The downside of making it a part of gradcheck right now is modifying a lot of callsites of gradcheck to not check this (because _batched_grad_grad_test fails for a lot of operators).

I think in the long run we should have gradcheck test for batched gradient, but it might take a while to get there

File: test/test_vmap.py [hunk 3]
```diff
@@ -640,35 +643,42 @@ def randn(size, device='cpu', dtype=torch.float):
...
+               check_view=False, check_propagates_grad=True):
+    result = vmap(op, in_dims, out_dims)(*inputs)
+    reference_result = reference_vmap(op, inputs, in_dims, out_dims)
+    self.assertEqual(result, reference_result)
+    op_has_single_return = not isinstance(result, tuple)
+
+    if check_view:
+        result_as_tuple = (result,) if op_has_single_return else result
+        for output in result_as_tuple:
+            self.assertEqual(output.data_ptr() - output.storage_offset() * output.element_size(),
```
Comment (albanD) [inline test/test_vmap.py:661]: Note that you could try as well to check that `output._base is inputs[0] if inputs[0]._base is None else inputs[0]._base` ?

File: test/test_vmap.py [hunk 4]
```diff
@@ -1150,5 +1160,100 @@ def test_no_random_op_support(self):
...
+    if istuple(x):
+        return x
+    elif isinstance(x, list):
+        return tuple(x)
+    else:
+        return x,
+
+def differentiable(args):
+    return tuple(arg for arg in as_tuple(args)
+                 if isinstance(arg, torch.Tensor) and arg.requires_grad)
```
Comment (albanD) [inline test/test_vmap.py:1177]: I never know which one is supposed to be better `isinstance(arg, torch.Tensor)` or `torch.is_tensor(arg)` ?

File: test/test_vmap.py [hunk 5]
```diff
@@ -1,8 +1,11 @@
 from torch.testing._internal.common_utils import TestCase, run_tests
 import torch
-from torch import vmap
+from torch import Tensor, vmap
 import functools
 import warnings
+from torch._six import istuple
```
[the remark this answers is not in the document]
Comment (zou3519) [reply test/test_vmap.py:6]: Good point. I was copy-pastaing some code

File: test/test_vmap.py (hunk 4 above)
Comment (zou3519) [reply test/test_vmap.py:1177]: https://github.com/pytorch/pytorch/blob/0ab83f7f9fcbbc9b7a709b99b38577d14c22a7b7/torch/__init__.py#L226-L231 We seem to recommend that users use `isinstance(arg, torch.Tensor)`.

File: test/test_vmap.py (hunk 3 above)
Comment (zou3519) [reply test/test_vmap.py:661]: I like that better, actually. I'll update the code to do that

File: test/test_vmap.py [hunk 6]
```diff
@@ -1150,5 +1159,100 @@ def test_no_random_op_support(self):
...
+    return torch.randn(batch_size, *output.shape,
+                       dtype=output.dtype, device=output.device)
+
+def as_tuple(x):
+    if isinstance(x, tuple):
+        return x
+    elif isinstance(x, list):
+        return tuple(x)
+    else:
+        return x,
```
Comment (ezyang) [inline test/test_vmap.py:1172]: No action really necessary here, but we have some other code like this in other places. It reminds me a bit of https://jax.readthedocs.io/en/latest/pytrees.html too

File: test/test_vmap.py [hunk 7]
```diff
@@ -1150,5 +1159,100 @@ def test_no_random_op_support(self):
...
+            assert len(outputs) > 0
+            return outputs
+
+        self._vmap_test(vector_hessian_product, batched_vectors,
+                        check_propagates_grad=False)
+
+    def test_sigmoid(self, device):
+        # Maybe we can make the "check that the slow fallback was not invoked"
+        # into a context manager, because it's used a lot. I'll leave that for
+        # future work.
```
Comment (ezyang) [inline test/test_vmap.py:1238]: Yes, the Not regex here looks very brittle; you won't notice it's out of date if you change the message

File: test/test_vmap.py [hunk 8]
```diff
@@ -1150,5 +1159,100 @@ def test_no_random_op_support(self):
...
+                        check_propagates_grad=False)
+
+    # Tests batched second grad computation of outputs = op(*args, **kwargs).
+    # by comparing it to a sequential map+stack fallback.
+    #
+    # output_process_fn: a function that maps the outputs to the part
+    #       that should be differentiated.
+    # batch_size: the batch dim size for the batched grad
+    #
+    # NB: we only test computing batched gradients are in the second gradient
```
Comment (ezyang) [inline test/test_vmap.py:1206]: grammar problem here

File: test/test_vmap.py [hunk 9]
```diff
@@ -1150,5 +1159,100 @@ def test_no_random_op_support(self):
...
+    # computation. One specific use case that does this is computing the hessian
+    # matrix of a scalar-valued function; this is useful in Bayesian Logistic
+    # Regression.
```
Comment (ezyang) [inline test/test_vmap.py:1209]: If you've got a citation, that would be a great thing to put here

File: test/test_vmap.py [hunk 10]
```diff
@@ -1150,5 +1159,100 @@ def test_no_random_op_support(self):
...
+    def _batched_grad_grad_test(self, op, args, kwargs, output_process_fn=lambda x: x, batch_size=3):
+        outputs = op(*args, **kwargs)
+        outputs = differentiable(output_process_fn(outputs))
+        ones = tuple(torch.ones_like(out) for out in outputs)
+        # Same thing as summing together all of the outputs and calling .backward()
+        first_grads = torch.autograd.grad(outputs, differentiable(args), ones,
+                                          create_graph=True)
+        first_grads = differentiable(first_grads)
+        if len(first_grads) == 0:
+            return  # None of the grads actually depend on the input
```
Comment (ezyang) [inline test/test_vmap.py:1221]: Maybe this should just fail here? It implies that you shouldn't have run this particular test on the operator in question--seems more likely to be a bug than intentional

Comment (zou3519) [reply test/test_vmap.py:1221]: yeah, that sounds reasonable

File: test/test_vmap.py (hunk 9 above)
Comment (zou3519) [reply test/test_vmap.py:1209]: No citation, I am just quoting something from someone

File: test/test_vmap.py (hunk 7 above)
Comment (zou3519) [reply test/test_vmap.py:1238]: Yup, I made it more structured in another PR (https://github.com/pytorch/pytorch/pull/43731)

File: test/test_vmap.py (hunk 2 above)
Comment (ezyang) [reply test/test_vmap.py:1212]: We have a sort of similar situation with complex numbers. My current thinking is that it's fine to not make it part of gradcheck in the beginning when it's only a handful of ops that support batched gradient. Then as you start adding support fo more operators, at some point you will get to the point where you are going for "completion"; at that point, it makes sense to flip the switch, and go from whitelisting ops which do work, to blacklisting ops that don't work.

**Review by ezyang — APPROVED** [review]
Sure. I cc'ed @mruberry in case he had more testing related thoughts

**Review by mruberry — APPROVED** [review]
Hey @zou3519, I wish I was more familiar with your vmap work (I hope to become more familiar with it soon!), but just from a test perspective this looks pretty good.
